Add multithreading support to subkey

chevdor opened the pull request:
This PR brings the following to `subkey`/vanity:
- ability to customize the specs for the key generation (case sensitive, minimum score, etc...)
- multithreading
- display of the results during the generation (in case you let subkey look for keys for hours...)
- display of the progress
- nicer display of the results
- sorting and display of the keys in a CSV friendly format

```
$ subkey vanity --help
subkey-vanity 
Generate vanity address

USAGE:
    subkey vanity [FLAGS] [OPTIONS] [pattern]

FLAGS:
    -c, --case-sensitive    Whether the patten case is important or not
    -h, --help              Prints help information
    -p, --paranoiac         Sacrifice speed for better randomness (~7% slower)
    -V, --version           Prints version information

OPTIONS:
    -f, --format <format>        Format of the generated output (stdout|csv) [default: stdout]
    -m, --minscore <minscore>    Minimum score for the results (between 0 and 100) [default: 75]
    -n, --number <number>        Number of keys to generate [default: 1]
    -t, --threads <threads>      Force the number of threads to be used (0= number of cpus x 2) [default: 0]

ARGS:
    <pattern>    Desired pattern
```

rphmeier commented on subkey/src/main.rs:
```diff
@@ -16,31 +22,50 @@
+				 case_sensitive: matches.is_present("case_sensitive"),
+				 paranoiac: matches.is_present("paranoiac"),
+				 minscore: match matches.value_of("minscore").map(str::to_string).unwrap_or_default().parse::<u8>().unwrap() {
+					 m @ 0...100 => m as f32,
```
extra space before `m` it seems

rphmeier commented on subkey/src/main.rs:
```diff
@@ -16,31 +22,50 @@
+				 paranoiac: matches.is_present("paranoiac"),
+				 minscore: match matches.value_of("minscore").map(str::to_string).unwrap_or_default().parse::<u8>().unwrap() {
+					 m @ 0...100 => m as f32,
+					_ => 100_f32,
```
perhaps this could be more efficiently written as `::std::cmp::min(m, 100) as f32`?

chevdor replied:
ok, fixed.

rphmeier commented on subkey/src/main.rs:
```diff
@@ -16,31 +22,50 @@
+				 desired_pattern: matches.value_of("pattern").map(str::to_string).unwrap_or_default(),
+				 case_sensitive: matches.is_present("case_sensitive"),
+				 paranoiac: matches.is_present("paranoiac"),
+				 minscore: match matches.value_of("minscore").map(str::to_string).unwrap_or_default().parse::<u8>().unwrap() {
```
`unwrap` not allowed, rather exit cleanly if it doesn't parse

rphmeier commented on subkey/src/vanity.rs:
```diff
@@ -36,67 +48,147 @@ fn next_seed(mut seed: [u8; 32]) -> [u8; 32] {
+pub fn generate_keys(
+	key_specs: KeySpecs,
+	count: usize,
+	threads: u8) -> Vec<KeyPair> {
```
close paren on next line

chevdor replied:
the multiline was historical...
fixed

rphmeier commented on subkey/src/vanity.rs:
```diff
@@ -36,67 +48,147 @@ fn next_seed(mut seed: [u8; 32]) -> [u8; 32] {
+	threads: u8) -> Vec<KeyPair> {
+
+    let (s, r) = channel();
+    let mut result: Vec<KeyPair>  = Vec::new();
```
mixed tabs and spaces in this file

chevdor replied:
ouch! fixed

chevdor commented:
Hi @gavofyork, thanks for the feedback.

> I already removed "paranoiac" because it's a pointless pretence at higher security that doesn't actually deliver any higher security.

It can be removed but I still see it in the code (https://github.com/paritytech/substrate/blob/master/subkey/src/vanity.rs#L72) and did not want to remove that in the PR and change the current behaviour. That's easy to remove though and I agree about the limited benefit. I will remove it.

> I already removed matching on partial finds because it's pointless and based on a misunderstanding on how scoring works. It's especially useless when given as a (entirely meaningless) percentage.

The motivation behind the percentage is to avoid using a non normalised value. How can users decide to use a threshold of 173 or 244 without knowing how the scoring works? At least the % helps getting a feel about what is going on.

> I already removed multi-key generation, because I see little utility in the context of vanity address generation and because it's trivial to call subkey multiple times from any shell script.

I would disagree about this one. If you do a training for a business customer, you may need N vanity addresses containing the name of the customer... This is non breaking as the default value is 1 anyway. There are only benefits. Moreover, scripting capabilities are really OS depend...

> Subkey already displays occasional progress reports.

True but with this PR, the `10000 keys processed` are now per thread. I left it as is so far but this is not very useful to the user. The idea behind the progress bar is mainly to know how far we got in the task of generating those N addresses the user requested. If I need 150 addresses, the amount of keys tested is cool to see but not really useful. I guess I could skip the progress bar if the request is limited to 1 key though. Would that be better?

gavofyork commented:
> The motivation behind the percentage is to avoid using a non normalised value. How can users decide to use a threshold of 173 or 244 without knowing how the scoring works? At least the % helps getting a feel about what is going on.

Scoring isn't meant as a metric for the user to set thresholds on at all; it's there only to help indicate how far through the expected process of finding a match it has come. Anything less than the final value must currently be understood to be a pointless result and the scoring system reengineered should it be used for such a purpose. Trying to fudge it into a percentage just lures the user into false sense of meaning.

> Moreover, scripting capabilities are really OS depend

As of right now, `substrate` compiles only for Mac and Linux, both of which come packaged with `bash`, which has `for` loops that make repetition trivial (`for i in {1..10}; do subkey myname; done`).

> in the task of generating those N addresses the user requested

I see "N address" generation this as the responsibility of shell scripting. There are no real gains by introducing this functionality into `subkey` and very real losses in terms of complexity and maintainability.

gavofyork commented:
Overall it seems that this PR is really trying to turn `subkey` into something that it really wasn't designed to be, so I feel a separate repo and project would be a perfectly reasonable solution.

chevdor commented:
If batch loops are an option, multithreading can be worked around as well with the current version with something like:
`for i in {0..8}; do subkey vanity abcde & done`.

The performance is not similar but "it works".

So I close this PR that does not seem to trigger much interest. The multithreaded version will remain available in my repo for people who cannot or do not want to use bash trickeries to generated keys or if the level of interest in that PR raises.
